File: py/kruskal.py
import logging
from typing import Optional

from util import Edge, Network, Vertice

logger = logging.getLogger('kruskal')


class Node:
    def __init__(self, vertice: Vertice) -> None:
        self.vertice = vertice
        self.parent: Optional[Node] = None
        self.rank = 0

    def __str__(self) -> str:
        if self.parent is not None:
            return f'{self.vertice}<-{self.parent.vertice}'
        else:
            return f'{self.vertice}<-.'

    def __repr__(self) -> str:
        return str(self.vertice)


class KruskalEdge:
    def __init__(self, u: Node, v: Node, edge: Edge) -> None:
        self.raw_edge = edge
        self.u = u
        self.v = v
        self.cost = edge.cost

    def __str__(self) -> str:
        return f'{self.u!r}<-{self.cost}->{self.v!r}'


def make_set(vertice: Vertice) -> Node:
    return Node(vertice)


def find(x: Node) -> Node:
    if x.parent is not None:
        x.parent = find(x.parent)
        return x.parent
    else:
        return x


def union(x: Node, y: Node):
    x_root = find(x)
    y_root = find(y)

    if x_root != y_root:
        if x_root.rank < y_root.rank:
            x_root.parent = y_root
        else:
            y_root.parent = x_root
            if x_root.rank == y_root.rank:
                x_root.rank += 1


def solve(network: Network) -> Network:
    log = logger.getChild(solve.__name__)
    nodes = {vertice: make_set(vertice) for vertice in network.vertices_sorted()}
    edges = [KruskalEdge(nodes[edge.u], nodes[edge.v], edge) for edge in network.edges_sorted()]
    edges = sorted(edges, key=lambda edge: edge.cost)

    tree = Network()

    def repr_nodes(nodes: dict[Node]) -> str:
        return ', '.join([str(v) for v in nodes.values()])

    log.debug(f'len_nodes={len(nodes):2}, nodes={nodes}')
    log.debug(f'len_edges={len(edges):2}, edges={edges!r}')
    for edge in edges:
        log.debug(f'current edge: {edge}')
        log.debug(f'nodes: {repr_nodes(nodes)}')
        if find(edge.u) != find(edge.v):
            tree.insert_edge(edge.raw_edge)
            union(edge.u, edge.v)
        else:
            log.debug(f'drop link {edge.u.vertice}<->{edge.v.vertice}')

    return tree

File: src/kruskal.rs
use std::{cell::RefCell, rc::Rc};

use crate::solv::{Edge, Network, Solver, VerticeId};

type WrappedNode = Rc<RefCell<Node>>;
pub struct Kruskal {}

impl Solver for Kruskal {
    fn solve(network: Network) -> Network {
        let vertices = network.vertices_sorted();
        let nodes = transform_vertices(vertices);
        let edges = transform_edges(network.edges_sorted(), &nodes);

        let mut tree = Network::default();

        for edge in edges {
            let u = &edge.u;
            let v = &edge.v;
            if find(u) != find(v) {
                tree.insert_edge(edge.raw_edge);
                union(u, v);
            } else {
                log::debug!("drop edge {}", edge.raw_edge);
            }
        }

        tree
    }
}

fn transform_vertices(vertices: Vec<VerticeId>) -> Vec<WrappedNode> {
    vertices.into_iter().map(make_set).collect()
}

fn transform_edges(edges: Vec<Edge>, nodes: &[WrappedNode]) -> Vec<NoddedEdge> {
    let mut result: Vec<NoddedEdge> = edges
        .into_iter()
        .map(|edge| NoddedEdge::from_edge(edge, nodes))
        .collect();
    result.sort_by(|a, b| a.weight.cmp(&b.weight));
    result
}

struct NoddedEdge {
    pub u: WrappedNode,
    pub v: WrappedNode,
    pub raw_edge: Edge,
    pub weight: usize,
}

impl NoddedEdge {
    fn from_edge(edge: Edge, available_nodes: &[WrappedNode]) -> Self {
        let mut nodes_it = available_nodes.iter();
        let u = nodes_it
            .find(|node| node.borrow().vertice == edge.u)
            .expect("cannot find vertice u")
            .clone();
        let v = nodes_it
            .find(|node| node.borrow().vertice == edge.v)
            .expect("cannot find vertice v")
            .clone();

        Self {
            u,
            v,
            raw_edge: edge.clone(),
            weight: edge.weight,
        }
    }
}

#[derive(PartialEq, Eq)]
struct Node {
    vertice: VerticeId,
    parent: Option<WrappedNode>,
    rank: usize,
}

impl Node {
    fn new(vertice: VerticeId) -> Self {
        Self {
            vertice,
            parent: None,
            rank: 0,
        }
    }
}

fn find(x: &WrappedNode) -> WrappedNode {
    x.borrow()
        .parent
        .as_ref()
        .map_or_else(|| x.clone(), |parent| parent.clone())
}

fn make_set(vertice: VerticeId) -> WrappedNode {
    Rc::new(RefCell::new(Node::new(vertice)))
}

fn union(x: &WrappedNode, y: &WrappedNode) {
    let x_root = find(x);
    let y_root = find(y);

    if x_root == y_root {
        return;
    }

    let mut x = x_root.borrow_mut();
    let mut y = y_root.borrow_mut();

    if x.rank < y.rank {
        std::mem::swap(&mut x, &mut y);
    }

    y.parent = Some(x_root.clone());
    if x.rank == y.rank {
        x.rank += 1;
    }
}
